### PR TITLE
Edge mesh input

### PR DESCRIPTION
**Pull request by Zhouyuan-Chen**
the first edge_mesh_input pr

### CODE REVIEW

**Comment by Zhouyuan-Chen** [comment]
I wrote a simple .obj EdgeMeshReader. can only support edges and vertices input.

**Comment by Zhouyuan-Chen** [comment]
I debugged the .obj file reading code. I wrote a sample .obj file under data/ called edge_test.obj please check if the format I read is correct.

File: src/wmtk/io/EdgeMeshReader.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+void EdgeMeshReader::read(Eigen::Matrix<long, -1, -1>& E, Eigen::MatrixXd& V)
+{
+    std::vector<std::pair<long, long>> edges;
+    std::vector<std::vector<double>> vertices;
+    char* line_buffer = nullptr;
+    std::fstream f(m_filename);
+    if (!f) {
+        throw std::runtime_error("can't open file!");
+    }
+    if (m_type == OBJ) {
```
Comment (daniel-zint) [inline src/wmtk/io/EdgeMeshReader.cpp:18]: use `switch` here instead of `if`

File: src/wmtk/io/EdgeMeshReader.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        std::string buffer;
```
Comment (daniel-zint) [inline src/wmtk/io/EdgeMeshReader.cpp:19]: There is a lot of code. You can push that into a local function to keep `read()` clean.

File: src/wmtk/io/EdgeMeshReader.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+EdgeMeshReader::EdgeMeshReader(const std::string& filename, const file_type type)
+    : m_filename(filename)
+    , m_type(type)
+{}
+void EdgeMeshReader::read(Eigen::Matrix<long, -1, -1>& E, Eigen::MatrixXd& V)
+{
+    std::vector<std::pair<long, long>> edges;
+    std::vector<std::vector<double>> vertices;
+    char* line_buffer = nullptr;
+    std::fstream f(m_filename);
```
Comment (daniel-zint) [inline src/wmtk/io/EdgeMeshReader.cpp:14]: use `std::ifstream`

File: src/wmtk/io/EdgeMeshReader.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    if (!f) {
```
Comment (daniel-zint) [inline src/wmtk/io/EdgeMeshReader.cpp:15]: `f.is_open()` might do the same thing (I am not sure) but it for sure makes your intend more clear.

File: src/wmtk/io/EdgeMeshReader.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        throw std::runtime_error("can't open file!");
+    }
+    if (m_type == OBJ) {
+        std::string buffer;
+        while (!f.eof()) {
+            getline(f, buffer);
+            // here is for debugging
+            // spdlog::info("lineinfo: {}", buffer[0]);
+            if (buffer[0] != 'v' && buffer[0] != 'l') continue;
```
Comment (daniel-zint) [inline src/wmtk/io/EdgeMeshReader.cpp:24]: No one line if-statements. Put brackets around `continue;`

File: src/wmtk/io/EdgeMeshReader.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            // read obj
+            line_buffer = new char[buffer.size()];
+            strcpy(line_buffer, buffer.c_str());
```
Comment (daniel-zint) [inline src/wmtk/io/EdgeMeshReader.cpp:27]: You can use a `std::string` just like an array. There is no need to copy the string into a char array.
For example, you can also just write `buffer[0] == 'v'`.

File: src/wmtk/io/EdgeMeshReader.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            if (line_buffer[0] == 'v') {
+                std::vector<double> positions;
+                char* token = strtok(line_buffer, " ");
```
Comment (daniel-zint) [inline src/wmtk/io/EdgeMeshReader.cpp:30]: Use a `std::istringstream` instead.

**Review by daniel-zint — COMMENTED** [review]
Only reviewing the `EdgeMeshReader`. While the concept seems to be fine, it is quite C heavy. Please try to use C++ `std` library functions instead.

**Comment by Zhouyuan-Chen** [comment]
this part code should be done if we only need one kind of file format.
wait to integrate the edgemesh part code into input.cpp.

**Pull request closed without merging**